sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/utils/CalciteUtils.java

```diff
@@ -38,149 +37,152 @@
  */
 public class CalciteUtils {
   private static final long UNLIMITED_ARRAY_SIZE = -1L;
-  private static final BiMap<SqlTypeCoder, SqlTypeName> BEAM_TO_CALCITE_TYPE_MAPPING =
-      ImmutableBiMap.<SqlTypeCoder, SqlTypeName>builder()
-          .put(SqlTypeCoders.TINYINT, SqlTypeName.TINYINT)
-          .put(SqlTypeCoders.SMALLINT, SqlTypeName.SMALLINT)
-          .put(SqlTypeCoders.INTEGER, SqlTypeName.INTEGER)
-          .put(SqlTypeCoders.BIGINT, SqlTypeName.BIGINT)
+  private static final BiMap<Schema.FieldType, SqlTypeName> BEAM_TO_CALCITE_TYPE_MAPPING =
+      ImmutableBiMap.<Schema.FieldType, SqlTypeName>builder()
+          .put(FieldType.BYTE, SqlTypeName.TINYINT)
+          .put(FieldType.INT16, SqlTypeName.SMALLINT)
+          .put(FieldType.INT32, SqlTypeName.INTEGER)
+          .put(FieldType.INT64, SqlTypeName.BIGINT)
 
-          .put(SqlTypeCoders.FLOAT, SqlTypeName.FLOAT)
-          .put(SqlTypeCoders.DOUBLE, SqlTypeName.DOUBLE)
+          .put(FieldType.FLOAT, SqlTypeName.FLOAT)
+          .put(FieldType.DOUBLE, SqlTypeName.DOUBLE)
 
-          .put(SqlTypeCoders.DECIMAL, SqlTypeName.DECIMAL)
+          .put(FieldType.DECIMAL, SqlTypeName.DECIMAL)
 
-          .put(SqlTypeCoders.CHAR, SqlTypeName.CHAR)
-          .put(SqlTypeCoders.VARCHAR, SqlTypeName.VARCHAR)
+          .put(FieldType.STRING, SqlTypeName.VARCHAR)
 
-          .put(SqlTypeCoders.DATE, SqlTypeName.DATE)
-          .put(SqlTypeCoders.TIME, SqlTypeName.TIME)
-          .put(SqlTypeCoders.TIMESTAMP, SqlTypeName.TIMESTAMP)
+          .put(FieldType.DATETIME, SqlTypeName.TIMESTAMP)
 
-          .put(SqlTypeCoders.BOOLEAN, SqlTypeName.BOOLEAN)
-          .build();
+          .put(FieldType.BOOLEAN, SqlTypeName.BOOLEAN)
 
-  private static final BiMap<SqlTypeName, SqlTypeCoder> CALCITE_TO_BEAM_TYPE_MAPPING =
+          .put(FieldType.ARRAY, SqlTypeName.ARRAY)
+          .put(FieldType.ROW, SqlTypeName.ROW)
```

Comment:
<!--thread_id:cc_178369253_t; commit:79c95678e593da730ba0472b77304ec1f916245e; resolved:0-->
<!--section:context-quote-->
> **akedin** wrote:
> Not sure if this will work, but I would try to change this map to `BiMap<FieldTypeDescriptor, SqlTypeName>`, so that it is possible to do it like this:
> 
> ```java
>   BEAM_TO_CALCITE_MAPPING = 
>     ...
>     .put(FieldType.DECIMAL.typeDescriptor(), SqlTypeName.DECIMAL)
>     .put(FieldType.STRING.typeDescriptor().withMetadata(CHAR_METADATA), SqlTypeName.CHAR)
>     .put(FieldType.STRING.typeDescriptor().withMetadata(VARCHAR_METADATA), SqlTypeName.VARCHAR)
>     .put(FieldType.DATETIME.typeDescriptor().withMetadata(TIME_METADATA), SqlTypeName.TIME)
>     .put(FieldType.DATETIME.typeDescriptor().withMetadata(TIMESTAMP_METADATA), SqlTypeName.TIMESTAMP)
>      ...
> ```
> 
> Then you probably don't need to look up metadata separately each time and now have a 1-1 mapping

<!--section:body-->
Good idea: done. Required a bit more work since we need to support mapping beam type -> sql type when the metadata isn't set (in this case we have a default mapping - e.g. String -> VARCHAR).